src/devices/src/virtio/persist.rs

```diff
@@ -0,0 +1,141 @@
+// Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+//! Defines the structures needed for saving/restoring Virtio primitives.
+
+use super::device::*;
+use super::queue::*;
+use crate::vm_memory::Address;
+use snapshot::Persist;
+use versionize::{VersionMap, Versionize, VersionizeResult};
+use versionize_derive::Versionize;
+use vm_memory::GuestAddress;
+
+use std::num::Wrapping;
+use std::sync::atomic::Ordering;
+
+#[derive(Clone, Debug, PartialEq, Versionize)]
+pub struct QueueState {
+    /// The maximal size in elements offered by the device
+    max_size: u16,
+
+    /// The queue size in elements the driver selected
+    size: u16,
+
+    /// Indicates if the queue is finished with configuration
+    ready: bool,
+
+    /// Guest physical address of the descriptor table
+    desc_table: u64,
+
+    /// Guest physical address of the available ring
+    avail_ring: u64,
+
+    /// Guest physical address of the used ring
+    used_ring: u64,
+
+    next_avail: Wrapping<u16>,
+    next_used: Wrapping<u16>,
+}
+
+impl Persist for Queue {
+    type State = QueueState;
+    type ConstructorArgs = ();
+    type Error = ();
+
+    fn save(&self) -> Self::State {
+        QueueState {
+            max_size: self.max_size,
+            size: self.size,
+            ready: self.ready,
+            desc_table: self.desc_table.0,
+            avail_ring: self.avail_ring.0,
+            used_ring: self.used_ring.0,
+            next_avail: self.next_avail,
+            next_used: self.next_used,
+        }
+    }
+
+    fn restore(
+        _: Self::ConstructorArgs,
+        state: &Self::State,
+    ) -> std::result::Result<Self, Self::Error> {
+        Ok(Queue {
+            max_size: state.max_size,
+            size: state.size,
+            ready: state.ready,
+            desc_table: GuestAddress::new(state.desc_table),
+            avail_ring: GuestAddress::new(state.avail_ring),
+            used_ring: GuestAddress::new(state.used_ring),
+            next_avail: state.next_avail,
+            next_used: state.next_used,
+        })
+    }
+}
+
+/// State of a VirtioDevice.
+#[derive(Debug, PartialEq, Versionize)]
+pub struct VirtioDeviceState {
+    pub avail_features: u64,
+    pub acked_features: u64,
+    pub device_type: u32,
+    pub queues: Vec<QueueState>,
+    pub interrupt_status: usize,
+    pub activated: bool,
+}
+
+impl VirtioDeviceState {
+    pub fn from_device(device: &dyn VirtioDevice) -> Self {
```

Comment:
Can we also implement Persist here? Personally I do not like the current solution as it looks like half the interface. Can we work around the usage of a Trait object so we can use Associated types ?

Reply:
I think that would imply making `VirtioDeviceState<T: VirtioDevice>`. I'm not sure it's nicer.

Since we don't have a clear alternative here, I suggest we move forward with this one. It can be changed down the line when someone brings a better alternative (it's a two-way door, does not impose constraints on format).

What do you think?

Reply:
It makes sense to go ahead with the current impl and rework this later.